components/core/src/clp/CurlDownloadHandler.cpp

```diff
@@ -48,13 +58,47 @@ CurlDownloadHandler::CurlDownloadHandler(
     m_easy_handle.set_option(CURLOPT_TIMEOUT, static_cast<long>(overall_timeout.count()));
 
     // Set up http headers
+    constexpr std::string_view cRangeHeaderName{"range"};
+    constexpr std::string_view cCacheControlHeaderName{"cache-control"};
+    constexpr std::string_view cPragmaHeaderName{"pragma"};
+    std::unordered_set<std::string_view> reserved_headers;
+    reserved_headers.insert(cRangeHeaderName);
+    reserved_headers.insert(cCacheControlHeaderName);
+    reserved_headers.insert(cPragmaHeaderName);
     if (0 != offset) {
-        std::string const range{"Range: bytes=" + std::to_string(offset) + "-"};
-        m_http_headers.append(range);
+        m_http_headers.append(fmt::format("{}: bytes={}-", cRangeHeaderName, offset));
     }
     if (disable_caching) {
-        m_http_headers.append("Cache-Control: no-cache");
-        m_http_headers.append("Pragma: no-cache");
+        m_http_headers.append(fmt::format("{}: no-cache", cCacheControlHeaderName));
+        m_http_headers.append(fmt::format("{}: no-cache", cPragmaHeaderName));
+    }
+    if (http_header_kv_pairs.has_value()) {
+        for (auto const& [key, value] : http_header_kv_pairs.value()) {
+            // Convert to lowercase for case-insensitive comparison
+            std::string lower_key = key;
+            std::transform(
+                    lower_key.begin(),
+                    lower_key.end(),
+                    lower_key.begin(),
+                    [](unsigned char c) { return std::tolower(c); }
```

Comment:
```suggestion
                    [](unsigned char c) -> char {
                        // Implicitly cast the input character into `unsigned char` to avoid UB:
                        // https://en.cppreference.com/w/cpp/string/byte/tolower
                        return static_cast<char>(std::tolower(c));
                    }
```
Let's add a comment to explain why we need `unsigned char`. Ideally we should move this to string_utils, but I'm fine to move it to future PRs.